File: Vanilla/sources/zomboid/zombie/iso/IsoUtils.java
package zombie.iso;

import org.joml.Vector2f;
import zombie.characters.IsoGameCharacter;
import zombie.characters.IsoPlayer;
import zombie.core.Core;

public final class IsoUtils {
   public static float clamp(float var0, float var1, float var2) {
      return Math.min(Math.max(var0, var1), var2);
   }

   public static float smoothstep(float var0, float var1, float var2) {
      float var3 = clamp((var2 - var0) / (var1 - var0), 0.0F, 1.0F);
      return var3 * var3 * (3.0F - 2.0F * var3);
   }

   public static float DistanceTo(float var0, float var1, float var2, float var3) {
      return (float)Math.sqrt(Math.pow((double)(var2 - var0), 2.0D) + Math.pow((double)(var3 - var1), 2.0D));
   }

   public static float DistanceTo2D(float var0, float var1, float var2, float var3) {
      return (float)Math.sqrt(Math.pow((double)(var2 - var0), 2.0D) + Math.pow((double)(var3 - var1), 2.0D));
   }

   public static float DistanceTo(float var0, float var1, float var2, float var3, float var4, float var5) {
      return (float)Math.sqrt(Math.pow((double)(var3 - var0), 2.0D) + Math.pow((double)(var4 - var1), 2.0D) + Math.pow((double)(var5 - var2), 2.0D));
   }

   public static float DistanceToSquared(float var0, float var1, float var2, float var3, float var4, float var5) {
      return (float)(Math.pow((double)(var3 - var0), 2.0D) + Math.pow((double)(var4 - var1), 2.0D) + Math.pow((double)(var5 - var2), 2.0D));
   }

   public static float DistanceToSquared(float var0, float var1, float var2, float var3) {
      return (float)(Math.pow((double)(var2 - var0), 2.0D) + Math.pow((double)(var3 - var1), 2.0D));
   }

   public static float DistanceManhatten(float var0, float var1, float var2, float var3) {
      return Math.abs(var2 - var0) + Math.abs(var3 - var1);
   }

   public static float DistanceManhatten(float var0, float var1, float var2, float var3, float var4, float var5) {
      return Math.abs(var2 - var0) + Math.abs(var3 - var1) + Math.abs(var5 - var4) * 2.0F;
   }

   public static float DistanceManhattenSquare(float var0, float var1, float var2, float var3) {
      return Math.max(Math.abs(var2 - var0), Math.abs(var3 - var1));
   }

   public static float XToIso(float var0, float var1, float var2) {
      float var3 = var0 + IsoCamera.getOffX();
      float var4 = var1 + IsoCamera.getOffY();
      float var5 = (var3 + 2.0F * var4) / (64.0F * (float)Core.TileScale);
      float var6 = (var3 - 2.0F * var4) / (-64.0F * (float)Core.TileScale);
      var5 += 3.0F * var2;
      float var10000 = var6 + 3.0F * var2;
      return var5;
   }

   public static float XToIsoTrue(float var0, float var1, int var2) {
      float var3 = var0 + (float)((int)IsoCamera.cameras[IsoPlayer.getPlayerIndex()].OffX);
      float var4 = var1 + (float)((int)IsoCamera.cameras[IsoPlayer.getPlayerIndex()].OffY);
      float var5 = (var3 + 2.0F * var4) / (64.0F * (float)Core.TileScale);
      float var6 = (var3 - 2.0F * var4) / (-64.0F * (float)Core.TileScale);
      var5 += (float)(3 * var2);
      float var10000 = var6 + (float)(3 * var2);
      return var5;
   }

   public static float XToScreen(float var0, float var1, float var2, int var3) {
      float var4 = 0.0F;
      var4 += var0 * (float)(32 * Core.TileScale);
      var4 -= var1 * (float)(32 * Core.TileScale);
      return var4;
   }

   public static float XToScreenInt(int var0, int var1, int var2, int var3) {
      return XToScreen((float)var0, (float)var1, (float)var2, var3);
   }

   public static float YToScreenExact(float var0, float var1, float var2, int var3) {
      float var4 = YToScreen(var0, var1, var2, var3);
      var4 -= IsoCamera.getOffY();
      return var4;
   }

   public static float XToScreenExact(float var0, float var1, float var2, int var3) {
      float var4 = XToScreen(var0, var1, var2, var3);
      var4 -= IsoCamera.getOffX();
      return var4;
   }

   public static float YToIso(float var0, float var1, float var2) {
      float var3 = var0 + IsoCamera.getOffX();
      float var4 = var1 + IsoCamera.getOffY();
      float var5 = (var3 + 2.0F * var4) / (64.0F * (float)Core.TileScale);
      float var6 = (var3 - 2.0F * var4) / (-64.0F * (float)Core.TileScale);
      float var10000 = var5 + 3.0F * var2;
      var6 += 3.0F * var2;
      return var6;
   }

   public static float YToScreen(float var0, float var1, float var2, int var3) {
      float var4 = 0.0F;
      var4 += var1 * (float)(16 * Core.TileScale);
      var4 += var0 * (float)(16 * Core.TileScale);
      var4 += ((float)var3 - var2) * (float)(96 * Core.TileScale);
      return var4;
   }

   public static float YToScreenInt(int var0, int var1, int var2, int var3) {
      return YToScreen((float)var0, (float)var1, (float)var2, var3);
   }

   public static boolean isSimilarDirection(IsoGameCharacter var0, float var1, float var2, float var3, float var4, float var5) {
      Vector2f var6 = new Vector2f(var1 - var0.x, var2 - var0.y);
      var6.normalize();
      Vector2f var7 = new Vector2f(var0.x - var3, var0.y - var4);
      var7.normalize();
      var6.add(var7);
      return var6.length() < var5;
   }
}

File: Vanilla/sources/zdoc/media/lua/shared/Library/IsoUtils.lua
---@class IsoUtils : zombie.iso.IsoUtils
IsoUtils = {}

---@public
---@param fromX float
---@param fromY float
---@param toX float
---@param toY float
---@return float
---@overload fun(fromX:float, fromY:float, toX:float, toY:float, fromZ:float, toZ:float)
function IsoUtils:DistanceManhatten(fromX, fromY, toX, toY) end

---@public
---@param fromX float
---@param fromY float
---@param toX float
---@param toY float
---@param fromZ float
---@param toZ float
---@return float
function IsoUtils:DistanceManhatten(fromX, fromY, toX, toY, fromZ, toZ) end

---@public
---@param objectX int
---@param objectY int
---@param objectZ int
---@param screenZ int
---@return float
function IsoUtils:YToScreenInt(objectX, objectY, objectZ, screenZ) end

---@public
---@param screenX float
---@param screenY float
---@param floor float
---@return float
function IsoUtils:XToIso(screenX, screenY, floor) end

---@public
---@param screenX float
---@param screenY float
---@param floor float
---@return float
function IsoUtils:YToIso(screenX, screenY, floor) end

---@public
---@param fromX float
---@param fromY float
---@param toX float
---@param toY float
---@return float
---@overload fun(arg0:float, arg1:float, arg2:float, arg3:float, arg4:float, arg5:float)
function IsoUtils:DistanceToSquared(fromX, fromY, toX, toY) end

---@public
---@param arg0 float
---@param arg1 float
---@param arg2 float
---@param arg3 float
---@param arg4 float
---@param arg5 float
---@return float
function IsoUtils:DistanceToSquared(arg0, arg1, arg2, arg3, arg4, arg5) end

---@public
---@param arg0 float
---@param arg1 float
---@param arg2 float
---@return float
function IsoUtils:smoothstep(arg0, arg1, arg2) end

---@public
---@param objectX float
---@param objectY float
---@param objectZ float
---@param screenZ int
---@return float
function IsoUtils:YToScreen(objectX, objectY, objectZ, screenZ) end

---@public
---@param fromX float
---@param fromY float
---@param toX float
---@param toY float
---@return float
---@overload fun(fromX:float, fromY:float, fromZ:float, toX:float, toY:float, toZ:float)
function IsoUtils:DistanceTo(fromX, fromY, toX, toY) end

---@public
---@param fromX float
---@param fromY float
---@param fromZ float
---@param toX float
---@param toY float
---@param toZ float
---@return float
function IsoUtils:DistanceTo(fromX, fromY, fromZ, toX, toY, toZ) end

---@public
---@param objectX float
---@param objectY float
---@param objectZ float
---@param screenZ int
---@return float
function IsoUtils:XToScreenExact(objectX, objectY, objectZ, screenZ) end

---@public
---@param objectX float
---@param objectY float
---@param objectZ float
---@param screenZ int
---@return float
function IsoUtils:YToScreenExact(objectX, objectY, objectZ, screenZ) end

---@public
---@param arg0 IsoGameCharacter
---@param arg1 float
---@param arg2 float
---@param arg3 float
---@param arg4 float
---@param arg5 float
---@return boolean
function IsoUtils:isSimilarDirection(arg0, arg1, arg2, arg3, arg4, arg5) end

---@public
---@param fromX float
---@param fromY float
---@param toX float
---@param toY float
---@return float
function IsoUtils:DistanceTo2D(fromX, fromY, toX, toY) end

---@public
---@param fromX float
---@param fromY float
---@param toX float
---@param toY float
---@return float
function IsoUtils:DistanceManhattenSquare(fromX, fromY, toX, toY) end

---@public
---@param screenX float
---@param screenY float
---@param floor int
---@return float
function IsoUtils:XToIsoTrue(screenX, screenY, floor) end

---@public
---@param objectX int
---@param objectY int
---@param objectZ int
---@param screenZ int
---@return float
function IsoUtils:XToScreenInt(objectX, objectY, objectZ, screenZ) end

---@public
---@param objectX float
---@param objectY float
---@param objectZ float
---@param screenZ int
---@return float
function IsoUtils:XToScreen(objectX, objectY, objectZ, screenZ) end

---@public
---@param arg0 float
---@param arg1 float
---@param arg2 float
---@return float
function IsoUtils:clamp(arg0, arg1, arg2) end
